crates/hc_sandbox/src/zome_call.rs

```diff
@@ -0,0 +1,516 @@
+    pub app_id: String,
+
+    /// The DNA hash to call
+    pub dna_hash: DnaHashB64,
+
+    /// The zome to call
+    pub zome_name: String,
+
+    /// The zome function to call
+    pub function: String,
+
+    /// The zome call payload as JSON
+    pub payload: String,
+}
+
+pub async fn zome_call_auth(
+    zome_call_auth: ZomeCallAuth,
+    admin_port: Option<u16>,
+) -> anyhow::Result<()> {
+    let admin_port = admin_port_from_connect_args(zome_call_auth.connect_args, admin_port).await?;
+
+    let app_client = AppClient::try_new(admin_port, zome_call_auth.app_id.clone()).await?;
+    let app_info = app_client.request(AppRequest::AppInfo).await?;
+    let info = match app_info {
+        AppResponse::AppInfo(Some(info)) => info,
+        other => anyhow::bail!("Unexpected response while getting app info: {:?}", other),
+    };
+
+    let cell_ids = info
+        .cell_info
+        .values()
+        .flatten()
+        .filter_map(|info| match info {
+            CellInfo::Provisioned(info) => Some(info.cell_id.clone()),
+            _ => None,
+        })
+        .collect::<HashSet<_>>();
+
+    let mut client = CmdRunner::try_new(admin_port).await?;
+
+    holochain_util::pw::pw_set_piped(zome_call_auth.piped);
+    if !zome_call_auth.piped {
+        msg!("Enter new passphrase to authorize zome calls: ");
+    }
+    let passphrase = holochain_util::pw::pw_get().context("Failed to get passphrase")?;
+
+    let (auth, key) = generate_signing_credentials(passphrase)?;
+
+    let signing_agent_key = AgentPubKey::from_raw_32(key.verifying_key().as_bytes().to_vec());
+
+    for cell_id in cell_ids {
+        client
+            .command(AdminRequest::GrantZomeCallCapability(Box::new(
+                GrantZomeCallCapabilityPayload {
+                    cell_id: cell_id.clone(),
+                    cap_grant: ZomeCallCapGrant::new(
+                        "sandbox".to_string(),
+                        CapAccess::Assigned {
+                            secret: auth.cap_secret,
+                            assignees: vec![signing_agent_key.clone()].into_iter().collect(),
+                        },
+                        GrantedFunctions::All,
+                    ),
+                },
+            )))
+            .await?;
+
+        msg!("Authorized zome calls for cell: {:?}", cell_id);
+    }
+
+    Ok(())
+}
+
+pub async fn zome_call(zome_call: ZomeCall, admin_port: Option<u16>) -> anyhow::Result<()> {
+    let admin_port = admin_port_from_connect_args(zome_call.connect_args, admin_port).await?;
+
+    let client = AppClient::try_new(admin_port, zome_call.app_id.clone())
+        .await
+        .context("Could not create app client")?;
+
+    let app_info = client.request(AppRequest::AppInfo).await?;
+    let info = match app_info {
+        AppResponse::AppInfo(Some(info)) => info,
+        other => anyhow::bail!("Unexpected response while getting app info: {:?}", other),
+    };
+
+    let cell_ids = info
```

Comment:
Interestingly, I noticed while testing it that the test_wasm_foo happ actually has two roles with identical cell_ids. I didn't even know that's allowed at all. I discovered that when trying to make sense of the following error that I saw in the holochain logs when making a zome call:
```
ERROR holochain_zome_types::dna_def: crates/holochain_zome_types/src/dna_def.rs:153: ZomeNotFound: zome1. (get_coordinator_zome) Existing zomes: integrity=[(ZomeName("zome1"), IntegrityZomeDef(Wasm(WasmZome { wasm_hash: WasmHash(uhCokg4OxLPjrBhP805xMp5d2kYn4_4HnfpivKyomxxDuMtBx4i_Y), dependencies: [], preserialized_path: None })))], coordinator=[]
```
The zome call succeeded anyway which still confuses me.

Reply:
Yep, it's a weird artifact of how zome calls are implemented, they will try more than one place looking for the right place to invoke the function and you get that error. Jost has pointed it out and it's something we should take a closer look at